### DISCUSSION TITLE
AWS ALB Authentication and next-auth

### DISCUSSION THREAD

**Discussion by zomgbre** [nextauthjs/next-auth#9781] [Help]
### Question 💬

Hi,

Is there any guidance on leveraging next-auth with AWS ALB authentication feature? 

https://docs.aws.amazon.com/elasticloadbalancing/latest/application/listener-authenticate-users.html
https://aws.amazon.com/blogs/aws/built-in-authentication-in-alb/

We are attempting to leverage this with a Nextjs 13 application and curious how one would go about doing so with this library or if it is even recommended. At the moment, we have implemented the verification steps documented on AWS documentation as middleware, however I like the additional functionality that next-auth brings with CSRF etc. Our team is new to Nextjs in general so any help greatly appreciated.

I know we could switch to a provider, but still curious if there is support for this method. :)

Thanks! ❤️ 

### How to reproduce ☕️

Here is the middleware we got so far - work in progress and not accessible on the public internet:
```
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';
import { logger } from './logger'
import * as jose from 'jose';

export const config = {
    matcher: '/((?!api\/health).*)',
};

export async function middleware(request: NextRequest) {
    const env = process.env.NODE_ENV;

    if (env == 'production') {
        const encodedJWTToken = request.headers.get('x-amzn-oidc-data');
        if (encodedJWTToken == undefined) return unAuthorizedResponse();

        try {
            const encodedJwtHeaders = encodedJWTToken.split('.')[0];
            const decodedJwtHeaders = Buffer.from(encodedJwtHeaders, 'base64').toString('utf-8');
            const jwtHeaders = JSON.parse(decodedJwtHeaders);
            const publicKeyResponse = await fetch(`https://public-keys.auth.elb.us-east-1.amazonaws.com/${jwtHeaders.kid}`);

            const publicKeyPem = await publicKeyResponse.text();
            const key = await jose.importSPKI(publicKeyPem, jwtHeaders.alg);
            const { payload } = await jose.jwtVerify(encodedJWTToken, key, { issuer: jwtHeaders.iss });

            logger.info({
                email: payload.email,
                url:  request.headers.get('referer')
            }, 'request recieved');

            return NextResponse.next();

        } catch (e) {
            return unAuthorizedResponse();
        }
    } else {
        // do nothing for local development :)
        return NextResponse.next();
    }
}

function unAuthorizedResponse() {
    return new NextResponse(
        JSON.stringify({ success: false, message: 'authentication failed' }),
        { status: 401, headers: { 'content-type': 'application/json' } }
    );
}
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by stale** [comment]
It looks like this issue did not receive any activity for 60 days. It will be closed in 7 days if no further activity occurs. If you think your issue is still relevant, commenting will keep it open. Thanks!

**Comment by oieduardorabelo** [comment]
Hey folks, any guidance on this topic? @balazsorban44 @ThangHuuVu
